inline_snapshot/_change.py

```diff
@@ -0,0 +1,222 @@
+import ast
+from collections import defaultdict
+from dataclasses import dataclass
+from typing import Any
+from typing import cast
+from typing import Dict
+from typing import List
+from typing import Optional
+from typing import Tuple
+from typing import Union
+
+from asttokens.util import Token
+from executing.executing import EnhancedAST
+from executing.executing import Source
+
+from ._rewrite_code import ChangeRecorder
+from ._rewrite_code import end_of
+from ._rewrite_code import start_of
+
+
+@dataclass()
+class Change:
+    flag: str
+    source: Source
+
+    @property
+    def filename(self):
+        return self.source.filename
+
+    def apply(self):
+        raise NotImplementedError()
+
+
+@dataclass()
+class Delete(Change):
+    node: ast.AST
+    old_value: Any
+
+
+@dataclass()
+class AddArgument(Change):
+    node: ast.Call
+
+    position: Optional[int]
+    name: Optional[str]
+
+    new_code: str
+    new_value: Any
+
+
+@dataclass()
+class ListInsert(Change):
+    node: ast.List
+    position: int
+
+    new_code: List[str]
+    new_values: List[Any]
+
+
+@dataclass()
+class DictInsert(Change):
+    node: ast.Dict
+    position: int
+
+    new_code: List[Tuple[str, str]]
+    new_values: List[Tuple[Any, Any]]
+
+
+@dataclass()
+class Replace(Change):
+    node: ast.AST
+
+    new_code: str
+    old_value: Any
+    new_value: Any
+
+    def apply(self):
+        change = ChangeRecorder.current.new_change()
+        range = self.source.asttokens().get_text_positions(self.node, False)
+        change.replace(range, self.new_code, filename=self.filename)
+
+
+TokenRange = Tuple[Token, Token]
+
+
+def generic_sequence_update(
+    source: Source,
+    parent: Union[ast.List, ast.Tuple, ast.Dict],
+    parent_elements: List[Union[TokenRange, None]],
+    to_insert: Dict[int, List[str]],
+):
+    rec = ChangeRecorder.current.new_change()
+
+    new_code = []
+    deleted = False
+    last_token, *_, end_token = source.asttokens().get_tokens(parent)
+    is_start = True
+    elements = 0
+
+    for index, entry in enumerate(parent_elements):
+        if index in to_insert:
+            new_code += to_insert[index]
+            print("insert", entry, new_code)
+        if entry is None:
+            deleted = True
+            print("delete1", entry)
```

Comment:
I noticed these prints while testing just now

Reply:
thank you. they are long gone in my local version.

Reply:
I pushed the new version. There should be no difference for you, except the print statements which are gone now.